File: spring-data-requery-java/src/test/java/com/coupang/springframework/data/requery/domain/sample/AbstractVersionedUser.java
package com.coupang.springframework.data.requery.domain.sample;

import com.coupang.kotlinx.objectx.ToStringBuilder;
import com.coupang.springframework.data.requery.domain.AbstractPersistable;
import io.requery.Entity;
import io.requery.Generated;
import io.requery.Key;
import io.requery.Version;
import org.jetbrains.annotations.NotNull;

import java.util.Date;
import java.util.Objects;

/**
 * AbstractVersionedUser
 *
 * @author debop@coupang.com
 * @since 18. 6. 25
 */
@Entity
public abstract class AbstractVersionedUser extends AbstractPersistable<Long> {

    @Key
    @Generated
    protected Long id;

    @Version
    protected Long version;


    protected String name;

    protected String email;

    protected Date birthday;

    public AbstractVersionedUser() {}

    public AbstractVersionedUser(String name, String email, Date birthday) {
        this.name = name;
        this.email = email;
        this.birthday = birthday;
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, email, birthday);
    }

    @Override
    protected @NotNull ToStringBuilder buildStringHelper() {
        return super.buildStringHelper()
            .add("name", name)
            .add("email", email)
            .add("birthday", birthday);
    }

    private static final long serialVersionUID = -7900602652543174851L;
}

File: src/test/kotlin/com/coupang/springframework/data/requery/domain/sample/AbstractVersionedUser.kt
package com.coupang.springframework.data.requery.domain.sample

import com.coupang.springframework.data.requery.domain.AbstractPersistable
import io.requery.*

/**
 * AbstractVersionedUser
 *
 * @author debop@coupang.com
 * @since 18. 5. 31
 */
@Entity
@Table(name = "Versioned_User")
abstract class AbstractVersionedUser: AbstractPersistable<Long>() {

    @get:Key
    @get:Generated
    abstract override val id: Long?

    @get:Column(name = "user_attr")
    abstract var attribute: String?

    @get:Column(name = "user_name")
    abstract var name: String?

    // NOTE: Superclass 인 경우에는 version 에 final 이 기본이지만, @Entity의 경우에는 `final`을 지정해주어야 합니다.
    @Version
    @JvmField
    final var version: Int = 0

}